hawc/apps/common/serializers.py

```diff
@@ -283,3 +286,119 @@ def run_validators(self, value):
     def create(self, validated_data, *args, **kwargs):
         instance, created = self.Meta.model.objects.get_or_create(**validated_data)
         return instance
+
+
+class BulkSerializer(serializers.ListSerializer):
+    """
+    Bulk create/update serializer for many = True
+    "id" must be given write access on serializer for bulk update
+    (eg.  id = IntegerField(required=False))
+    """
+
+    def to_internal_value(self, data):
+        """
+        List of dicts of native values <- List of dicts of primitive datatypes.
+        """
+        if html.is_html_input(data):
+            data = html.parse_html_list(data, default=[])
+
+        if not isinstance(data, list):
+            message = self.error_messages["not_a_list"].format(input_type=type(data).__name__)
+            raise ValidationError({api_settings.NON_FIELD_ERRORS_KEY: [message]}, code="not_a_list")
+
+        if not self.allow_empty and len(data) == 0:
+            message = self.error_messages["empty"]
+            raise ValidationError({api_settings.NON_FIELD_ERRORS_KEY: [message]}, code="empty")
+
+        ret = []
+        errors = []
+
+        for item in data:
+            try:
+                # https://github.com/miki725/django-rest-framework-bulk/issues/68#issuecomment-361895823
+                self.child.instance = (
+                    next((_ for _ in self.instance if _.id == item.get("id")), None)
+                    if self.instance
+                    else None
+                )
+                validated = self.child.run_validation(item)
+            except ValidationError as exc:
+                errors.append(exc.detail)
+            else:
+                ret.append(validated)
+                errors.append({})
+
+        if any(errors):
+            raise ValidationError(errors)
+
+        return ret
+
+    def validate_update(self, data):
+        # ids must be in data
+        data_ids = [obj_data.get("id") for obj_data in data if obj_data.get("id") is not None]
```

Comment:
The second check (all data ids should be in instance) is necessary for our bulk update to run without error. I could instead catch it within the `update` method, but I put it here instead so that all of our validation is in one place.

Reply:
oh yeah sorry, i wasn't clear. yes, please keep the second check, i'm just thinking you can simplify the first check and use the same data_ids in the second check.